File: CODES/Python/BinaryTree.py
class Node:
    def __init__(self, data):
        self.data = data
        self.left = None
        self.right = None

def create_node(data):
    return Node(data)

def insert_node(root, data):
    if root is None:
        return create_node(data)
    
    if data < root.data:
        root.left = insert_node(root.left, data)
    elif data > root.data:
        root.right = insert_node(root.right, data)
    
    return root

def inorder(root):
    if root is None:
        return
    inorder(root.left)
    print(root.data, end=" -> ")
    inorder(root.right)

def preorder(root):
    if root is None:
        return
    print(root.data, end=" -> ")
    preorder(root.left)
    preorder(root.right)

def postorder(root):
    if root is None:
        return
    postorder(root.left)
    postorder(root.right)
    print(root.data, end=" -> ")

if __name__ == "__main__":
    root = None
    root = insert_node(root, 50)
    insert_node(root, 30)
    insert_node(root, 20)
    insert_node(root, 40)
    insert_node(root, 70)
    insert_node(root, 60)
    insert_node(root, 80)

    print("Inorder traversal:")
    inorder(root)
    print("\n")

    print("Preorder traversal:")
    preorder(root)
    print("\n")

    print("Postorder traversal:")
    postorder(root)
    print("\n")

File: CODES/Javascript/binaryTree.js
class Node {
    constructor(data) {
        this.data = data;
        this.left = null;
        this.right = null;
    }
}

function createNode(data) {
    return new Node(data);
}

function insertNode(root, data) {
    if (root === null) {
        return createNode(data);
    }

    if (data < root.data) {
        root.left = insertNode(root.left, data);
    } else if (data > root.data) {
        root.right = insertNode(root.right, data);
    }

    return root;
}

function inorder(root) {
    if (root === null) {
        return;
    }

    inorder(root.left);
    console.log(root.data + " -> ");
    inorder(root.right);
}

function preorder(root) {
    if (root === null) {
        return;
    }

    console.log(root.data + " -> ");
    preorder(root.left);
    preorder(root.right);
}

function postorder(root) {
    if (root === null) {
        return;
    }

    postorder(root.left);
    postorder(root.right);
    console.log(root.data + " -> ");
}

let root = null;
root = insertNode(root, 50);
insertNode(root, 30);
insertNode(root, 20);
insertNode(root, 40);
insertNode(root, 70);
insertNode(root, 60);
insertNode(root, 80);

console.log("Inorder traversal:");
inorder(root);
console.log("\n");

console.log("Preorder traversal:");
preorder(root);
console.log("\n");

console.log("Postorder traversal:");
postorder(root);
console.log("\n");
